pylzy/lzy/api/v1/provisioning.py

```diff
@@ -16,46 +76,178 @@ class GpuType(Enum):
+
+
+@dataclasses.dataclass(frozen=True)
 class Provisioning:
-    cpu_type: Optional[str] = None
-    cpu_count: Optional[int] = None
-
-    gpu_type: Optional[str] = None
-    gpu_count: Optional[int] = None
-
-    ram_size_gb: Optional[int] = None
-
-    @staticmethod
-    def default() -> "Provisioning":
-        return Provisioning(
-            cpu_type=CpuType.CASCADE_LAKE.value,
-            cpu_count=2,
-            gpu_type=GpuType.NO_GPU.value,
-            gpu_count=0,
-            ram_size_gb=2,
-        )
+    cpu_type: StringRequirement = None
+    cpu_count: IntegerRequirement = None
+
+    gpu_type: StringRequirement = None
+    gpu_count: IntegerRequirement = None
+
+    ram_size_gb: IntegerRequirement = None
+
+    score_function: Optional[ScoreFunctionType] = dataclasses.field(
+        default=None,
+        repr=False,
+        compare=False,
+    )
+
+    proto_relation: ClassVar[Dict[str, str]] = {
+        'cpu_type': 'cpuType',
+        'cpu_count': 'cpuCount',
+        'gpu_type': 'gpuType',
+        'gpu_count': 'gpuCount',
+        'ram_size_gb': 'ramGb'
+    }
+
+    # NB: _final properties used only for type narrowing
 
-    def override(self, other: "Provisioning") -> "Provisioning":
-        return Provisioning(
-            cpu_type=other.cpu_type if other.cpu_type else self.cpu_type,
-            cpu_count=other.cpu_count if other.cpu_count else self.cpu_count,
-            gpu_type=other.gpu_type if other.gpu_type else self.gpu_type,
-            gpu_count=other.gpu_count if other.gpu_count else self.gpu_count,
-            ram_size_gb=other.ram_size_gb if other.ram_size_gb else self.ram_size_gb
+    @property
+    def cpu_type_final(self) -> str:
+        return _coerce(self.cpu_type, '')
+
+    @property
+    def gpu_type_final(self) -> str:
+        return _coerce(self.gpu_type, '')
+
+    @property
+    def cpu_count_final(self) -> int:
+        return _coerce(self.cpu_count, 0)
+
+    @property
+    def gpu_count_final(self) -> int:
+        return _coerce(self.gpu_count, 0)
+
+    @property
+    def ram_size_gb_final(self) -> int:
+        return _coerce(self.ram_size_gb, 0)
+
+    @property
+    def score_function_final(self) -> ScoreFunctionType:
+        return _coerce(self.score_function, minimum_score_function)
+
+    @classmethod
+    def from_proto(cls, spec: VmPoolSpec) -> "Provisioning":
+        fields = {}
+
+        for provisioning_field, spec_field in cls.proto_relation.items():
+            fields[provisioning_field] = getattr(spec, spec_field)
+
+        return cls(**fields)
+
+    def _filter_spec(self, spec: "Provisioning") -> bool:
+        if (
+            spec.cpu_count_final < self.cpu_count_final or
+            spec.gpu_count_final < self.gpu_count_final or
+            spec.ram_size_gb_final < self.ram_size_gb_final
+        ):
+            return False
+
+        if (
+            self.cpu_type_final and self.cpu_type_final != spec.cpu_type_final or
```

Comment:
I don't like checking for empty string here.
Maybe check if type is Any?

Reply:
Maybe return Union[str, AnyRequirement] from cpu_type_final, not just str?
It would be more obvious than just returning empty string as Any